Source/Scene/ModelExperimental/SceneMode2DPipelineStage.js

```diff
@@ -0,0 +1,284 @@
+import ModelExperimentalUtility from "./ModelExperimentalUtility.js";
+import SceneMode from "../SceneMode.js";
+import ShaderDestination from "../../Renderer/ShaderDestination.js";
+import VertexAttributeSemantic from "../VertexAttributeSemantic.js";
+import SceneTransforms from "../SceneTransforms.js";
+
+const scratchModelMatrix = new Matrix4();
+
+/**
+ * The scene mode 2D stage generates resources for rendering a primitive in 2D / CV mode.
+ *
+ * @namespace SceneMode2DPipelineStage
+ *
+ * @private
+ */
+const SceneMode2DPipelineStage = {};
+SceneMode2DPipelineStage.name = "SceneMode2DPipelineStage"; // Helps with debugging
+
+/**
+ * This pipeline stage processes the position attribute of a primitive and adds the relevant
+ * define and uniform matrix to the shader. It also generates new resources for the primitive
+ * in 2D. These resources persist in the runtime primitive so that the typed array used to
+ * store the positional data can be freed.
+ *
+ * This stage must go before the GeometryPipelineStage in the primitive pipeline.
+ *
+ * Processes a primitive. This stage modifies the following parts of the render resources:
+ * <ul>
+ *  <li> creates a vertex buffer for the positions of the primitive projected to 2D
+ *  <li> creates the bounding sphere for the primitive in 2D
+ *  <li> adds a flag to the shader to use 2D positions
+ *  <li> adds a uniform for the view model matrix in 2D
+ * </ul>
+ *
+ * @param {PrimitiveRenderResources} renderResources The render resources for this primitive.
+ * @param {ModelComponents.Primitive} primitive The primitive.
+ * @param {FrameState} frameState The frame state.
+ *
+ * @private
+ */
+
+SceneMode2DPipelineStage.process = function (
+  renderResources,
+  primitive,
+  frameState
+) {
+  const shaderBuilder = renderResources.shaderBuilder;
+  const runtimePrimitive = renderResources.runtimePrimitive;
+
+  const model = renderResources.model;
+  const modelMatrix = model.sceneGraph.computedModelMatrix;
+  const nodeComputedTransform = renderResources.runtimeNode.computedTransform;
+  const computedModelMatrix = Matrix4.multiplyTransformation(
+    modelMatrix,
+    nodeComputedTransform,
+    scratchModelMatrix
+  );
+
+  const boundingSphere2D = computeBoundingSphere2D(
+    renderResources,
+    computedModelMatrix,
+    frameState
+  );
+
+  runtimePrimitive.boundingSphere2D = boundingSphere2D;
+
+  const positionAttribute = ModelExperimentalUtility.getAttributeBySemantic(
+    primitive,
+    VertexAttributeSemantic.POSITION
+  );
+
+  // If the typed array of the position attribute exists, then
+  // the positions haven't been projected to 2D yet.
+  if (defined(positionAttribute.typedArray)) {
+    const buffer2D = createPositionBufferFor2D(
+      positionAttribute,
+      computedModelMatrix,
+      boundingSphere2D,
+      frameState
+    );
+
+    runtimePrimitive.positionBuffer2D = buffer2D;
+    model._modelResources.push(buffer2D);
+
+    // Unload the typed array
```

Comment:
Could we reword this to explain that when the typed array is shared, this doesn't cause issues because positionAttribute.typedArray is just a pointer to the array in the vertex buffer loader? This wasn't obvious.